src/lib.rs

```diff
@@ -32,9 +32,42 @@
 mod windows;
 
 use std::env;
+use std::ffi::OsString;
 use std::io;
 use std::path::{Path, PathBuf};
 
+/// Permits parameterising the home functions via the _from variants - used for
+/// in-process unit testing by rustup.
+pub trait Env {
+    fn home_dir(&self) -> Option<PathBuf>;
+    fn current_dir(&self) -> io::Result<PathBuf>;
+    fn var_os(&self, key: &str) -> Option<OsString>;
```

Comment:
Does `AsRef<OsStr>` bound to `key` make sense? As `std::env::var_os` has that too.

Reply:
No, this isn't env::var_os; we call it from limited call sites with static strings. There's no need for additional flexability today, and no reason to think we'll need it in future.

Reply:
`rustc_driver` has [`Callbakcs` trait](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/trait.Callbacks.html) which provides default implementation. And it doesn't provide DefaultCallbacks type which clippy has to define: https://github.com/rust-lang/rust-clippy/blob/c2c07fa9d095931eb5684a42942a7b573a0c5238/src/driver.rs#L66-L70.

Reply:
That's fair. We will change it if there are use cases in the future. For now, I agree to keep key as &str.

Reply:
> `rustc_driver` has [`Callbakcs` trait](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/trait.Callbacks.html) which provides default implementation. And it doesn't provide DefaultCallbacks type which clippy has to define: [https://github.com/rust-lang/rust-clippy/](https://github.com/rust-lang/rust-clippy/blob/c2c07fa9d095931eb5684a42942a7b573a0c5238/src/driver.rs#L66-L70)

Perhaps I misunderstand. We need to use the default implementation explicitly, because we are using the from variant. If it is not exposed we will have to implement the windows adaptation logic etc ourselves: at that point there is no value in using the home crate at all, and we may as well just inline it all into rustup. I linked to our thunk code.

Reply:
If we provide `Env` trait with default implementation, then in rustup we could juse redefined the OsEnv (or DefaultEnv)
and `impl home::Env for OsEnv`. This way we have default implementation like in "home" crate.

Reply:
I don't think a default implementation makes sense. Implemetors of the trait need to choose what behaviour they want, there is no semantically neutral behaviour.

Places where there is a sensible default are things like fast-path methods where a conservative default implementation can be written but a faster implementation can be written that takes advantage of the data structures internal knowledge.

Reply:
The [Callbacks trait][1] in rustc has sensible default I think.
Anyway, we already need that default in rustup.


[1]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_driver/trait.Callbacks.html

Reply:
If someone implements only half of those methods, the semantics of the trait are not violated. Thats different here, where the methods are a bundle. The only thing a default trait implementation would add is the ability for implementors to get it wrong.

Reply:
That's might be what users expect. I don't see anything wrong about that usecase.

Reply:
It's a foot gun, an attractive nuisance. Default implementations should only be used where the default is safe and reasonable, for implementors of the trait, not for receivers. Different audiences are involved.

Users that want the default behaviour here should not be calling the _from variants of the functions at all.

I don't understand why you're trying to keep the public symbol list smaller. What benefit does it bring?

Reply:
The only "reasonable" default implementations I can imagine would be to return `None` for  `var_os()` and `Err(NotImplemented)` for `current_dir()` which just moves the "oh I didn't implement" concern from compile time (where it's much easier to find) to runtime (where it might fester for ages before discovery), and perhaps to make `home_dir()` default to `self.var_os("HOME").map(PathBuf::from)`.  Better would be to offer the default implementation functions like `clap` does and like Robert has implemented here.  If you're trying to keep the API smaller then I'd eliminate all the convenience functions entirely and change the API to be:  `home::OS_ENV.home_dir()` etc since that makes it really clear that the home directory is coming from the OS environment.

Reply:
Ok, I gave up. I don't think I am worthy to maintain this crate.

I invited @kinnison as new owner of "home" and removed myself from "home" maintainers.

Reply:
This was certainly not my intent, I do not wish for you to feel driven off, nor unworthy.  You have been very reasonably questioning things so that we can clarify the intent of the patch.  However perhaps you and Robert have managed to talk past each other a little here.  Please don't give up on the crate because of me.

Reply:
No, I am sorry for being a bad gatekeeper.
Explaining/convincing has not always been my strong.

I just cannot convince myself to feel that this patch is in good state.
I don't want to be responsible to commit the changes and push it up
when I don't know that it is good.

After all, "home" was a glue between rustup and cargo (its plugins).
Having rustup devs to maintain it is better.

Good luck!

Reply:
I don't think you were being a bad gatekeeper, please don't take that away from this.  I think it's likely been a language/cross-purposes problem, but I certainly (and I imagine @rbtcollins too) value ensuring we're doing the right thing by communicating it effectively with the maintainers of a crate.  If you're adamant that you do not wish to continue maintaining `home` then assuming @brson is okay with it, we'll start the process of taking it in-board with the dev-tools team, but if you're prepared to continue as a maintainer then I value your input.  I'll accept the invitation on crates.io for now, but I don't know what the best approach is for deciding if/how to merge this PR and where to host the repo (Whether it should stay here or move) before any new release to crates.io is done.

Reply:
I too feel sad at this, @lzutao you are certainly worthy - and your contributions to rustup are well valued: I think you should consider yourself a rustup as much or more than I consider myself one!

I asked @kinnison to comment here because @lzutao and I were clearly not managing to communicate the ideas to each other, and I hoped @kinnison could find a different way of explaining us to each other.

I do *not* want to bully this patch in in bad state.

But also, I don't understand what is *bad* about it - I think the only remaining unaddressed concern you had was the existence and export of the OS environment struct.

I'd really appreciate it if you can help me understand what use case / behaviour / developer interaction is better with a default implementation. The fact that this matters to you means it is worth listening to for me and I presume @kinnison - we just need to find a way to articulate why this matters - because then we can talk about it.